Added rootca support

adamenger opened the pull request:
This attempts to address #1930. I have a root certificate at `/etc/vault/ssl/dev.pem` and I set the config via `fleetctl config set --rootca /etc/vault/ssl/dev.pem`. Once I do this, I'm able to login to my fleet instance which is fronted by a certificated generated through Vault:

```
➜  fleetctl config set --rootca /etc/vault/ssl/dev.pem
➜  ./build/fleetctl login
Log in using the standard Fleet credentials.
Email: aenger@etsy.com
Password:
[+] Fleet login successful and context configured!
```

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/kolide/fleet?pullRequest=1931) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/kolide/fleet?pullRequest=1931) before we can accept your contribution.<br/><hr/>**atom** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/kolide/fleet?pullRequest=1931) it.</sub>

zwass commented on cmd/fleetctl/api.go:
```diff
@@ -32,6 +35,10 @@ func unauthenticatedClientFromCLI(c *cli.Context) (*service.Client, error) {
 	}
 
+	if cc.RootCA != "" {
+		fleet = setupRootCA(fleet, cc.RootCA)
```
Instead of a separate function, let's do this root CA setup in service.NewClient.

zwass commented on cmd/fleetctl/api.go:
```diff
@@ -58,3 +65,26 @@ func clientFromCLI(c *cli.Context) (*service.Client, error) {
+	// read in the root cert file specified in the context
+	certs, err := ioutil.ReadFile(rootca)
+	if err != nil {
+		fmt.Println(fmt.Sprintf("Failed to append %v to RootCAs: %v", rootca, err))
```
fmt.Printf

zwass commented on cmd/fleetctl/config.go:
```diff
@@ -215,6 +221,13 @@ func configSetCommand() cli.Command {
+				EnvVar:      "ROOTCA",
+				Value:       "",
+				Destination: &flRootCA,
+				Usage:       "Root CA",
```
Let's be more clear about the purpose of this flag.

zwass commented on server/service/client.go:
```diff
@@ -16,7 +16,7 @@ type Client struct {
 	baseURL            *url.URL
 	token              string
-	http               *http.Client
+	Http               *http.Client
```
Moving setup to `service.NewClient` will allow us to leave this private.

zwass commented on cmd/fleetctl/api.go:
```diff
@@ -58,3 +65,26 @@ func clientFromCLI(c *cli.Context) (*service.Client, error) {
+
+func setupRootCA(client *service.Client, rootca string) *service.Client {
+	// get system pool, use empty on error
+	rootCAs, _ := x509.SystemCertPool()
```
Can we reason a bit more about whether this functionality should _append_ a cert to the system pool, or use _only_ the certs provided in the pem? I think it may be better to use only the provided cert (and error if this is missing) because it effectively allows pinning of the root.

adamenger commented:
Hi @zwass, i've addressed your comments and signed the CLA! Please let me know how this looks. One thing I'm unsure of is if my approach of adding another argument to `service.NewClient` is an acceptable approach. Will the rest of your invocations of that client in the codebase need to also pass in a rootCA now too? Also, would you like me to trap the error on line 35 in `server/service/client.go` ?

zwass commented on server/service/client.go:
```diff
@@ -30,9 +32,29 @@ func NewClient(addr string, insecureSkipVerify bool) (*Client, error) {
+		// read in the root cert file specified in the context
+		certs, err := ioutil.ReadFile(rootCA)
+		if err != nil {
+			fmt.Printf("Failed to append %v to RootCAs: %v\n", rootCA, err)
```
We need to propagate the error here, as otherwise we will continue with an empty cert chain. Please use `errors.Wrap` (many examples through this file).

zwass commented on server/service/client.go:
```diff
@@ -30,9 +32,29 @@ func NewClient(addr string, insecureSkipVerify bool) (*Client, error) {
+
+		// add certs to new cert pool
+		if ok := rootCAPool.AppendCertsFromPEM(certs); !ok {
+			fmt.Printf("No certs appended, using system certs only\n")
```
same here

adamenger commented:
hi @zwass I've added the `errors.Wrap()` and I also added some additional handling after line `35` like i mentioned previously just to be extra safe. Let me know what you think.

adamenger commented:
Ah I see CI failures, fixing...

adamenger commented:
Thank you for your guidance!
